### PR TITLE
Kademlia: Speed-up the record fetching

### PR DESCRIPTION
**Pull request by bkchr**
Before libp2p 0.50.0 we used a quorum of one to fetch records from the DHT. In the pr that upgraded to libp2p 0.50.0 we accidentally changed this behavior. This pr brings back the old behavior of using a qorum of one and thus, a faster discovery. After finding the first value, we directly finish the query. There was also another behavior change in libp2p, they stopped automatic caching on remote nodes. This pr also brings back the remote caching on nodes that are nearest to the key from our point of view of the network.

The pr that changed the behavior in libp2p: https://github.com/libp2p/rust-libp2p/pull/2712

### CODE REVIEW

File: client/network/src/discovery.rs [hunk 1]
```diff
@@ -692,33 +696,54 @@ impl NetworkBehaviour for DiscoveryBehaviour {
...
+								}
+
+								// Put the record to the `cache_candidates` that are nearest to the
+								// record key from our point of view of the network.
+								if let Some(record) = self.records_to_publish.remove(&id) {
+									if let Some(kad) = self.kademlia.as_mut() {
+										kad.put_record_to(
+											record,
+											cache_candidates.into_iter().map(|v| v.1),
+											Quorum::One,
```
Comment (dmitry-markin) [inline client/network/src/discovery.rs:740]: It's not clear from the docs: does this mean _only one_ peer will be updated, or _at least one_?

**Review by dmitry-markin — APPROVED** [review]
LGTM!

**Review by ggwpez — APPROVED** [review]
I dont know the code but it fixes the `zombienet-tests-misc-paritydb` test in 30s here: https://github.com/paritytech/polkadot/pull/6018

File: client/network/src/discovery.rs (hunk 1 above)
Comment (bkchr) [reply client/network/src/discovery.rs:740]: I think at least one.

**Review by melekes — COMMENTED** [review]
thanks for fixing these 🙏

File: client/network/src/discovery.rs [hunk 2]
```diff
@@ -692,33 +696,54 @@ impl NetworkBehaviour for DiscoveryBehaviour {
...
+
+								DiscoveryOut::ValueFound(
+									vec![(r.record.key, r.record.value)],
+									stats.duration().unwrap_or_default(),
+								)
+							},
+							Ok(GetRecordOk::FinishedWithNoAdditionalRecord {
+								cache_candidates,
+							}) => {
+								if cache_candidates.is_empty() {
```
Comment (skunert) [inline client/network/src/discovery.rs:729]: @bkchr A bit late to this party, but should we not call `self.records_to_publish.remove(&id)` here? 
I think if `cache_candidates` is empty, we could still have added something to `records_to_publish` for this query id which would not be cleared then.

Comment (bkchr) [reply client/network/src/discovery.rs:729]: :see_no_evil:

Comment (bkchr) [reply client/network/src/discovery.rs:729]: https://github.com/paritytech/substrate/pull/13086

**Pull request merged**